### PR TITLE
Alternative base for EncodedNumber

### PR DESCRIPTION
**Pull request by hardbyte**
A small code change, tests and docs.

### CODE REVIEW

File: docs/caveats.rst [hunk 1]
```diff
@@ -16,8 +15,27 @@ all exponents as part of the protocol; then for each
...
+
+    class AltEncodedNumber(paillier.EncodedNumber):
+        BASE = 2
+        LOG2_BASE = math.log(BASE, 2)
+
+
+.. warning::
+
+    Be careful when designing protocols with `BASE = 2`. If you don't require specific a
+    specific value for the unencrypted exponents after an operation then you're potentially
```
Comment (felixlawrence) [inline docs/caveats.rst]: typo: "specific a specific"

File: docs/caveats.rst [hunk 2]
```diff
@@ -16,8 +15,27 @@ all exponents as part of the protocol; then for each
...
+    leaking a lot of information about what happened.
```
Comment (felixlawrence) [inline docs/caveats.rst]: maybe I'd phrase this:
"As always, if you don't require a specific value for the unencrypted exponents after an operation, then might be leaking information about what happened - but with small bases this problem is exacerbated."

File: phe/tests/paillier_test.py [hunk 3]
```diff
@@ -193,63 +198,96 @@ def testEncodeIntDecodeInt4(self):
...
-        dec = paillier.EncodedNumber.BASE ** enc.exponent * enc.encoding
+        enc = self.EncodedNumberCls.encode(self.public_key, 15.1)
+        negexp = self.EncodedNumberCls.BASE ** enc.exponent
+        dec = self.EncodedNumberCls.BASE ** enc.exponent * enc.encoding
         self.assertEqual(15.1, dec)
 
     def testEncodeFloatDecodeFloat0(self):
-        enc = paillier.EncodedNumber.encode(self.public_key, 15.1)
+        enc = self.EncodedNumberCls.encode(self.public_key, 15.1)
         self.assertEqual(15.1, enc.decode())
```
Comment (felixlawrence) [inline phe/tests/paillier_test.py]: Might want to check whether this still works when BASE != 2**n... might have to slip in an `assertAlmostEqual` if you want to allow arbitrary BASEs

**Comment by hardbyte** [comment]
Ok thanks for the feedback @felixlawrence, I've added a test with an odd base and changed to `assertAlmostEquals` where necessary.

**Pull request closed without merging**